Add a performance measuring top-level user guide page (#10539)
haskell/cabal: doc/how-to-analyze-haskell-code-performance.rst

How to analyze Haskell performance
==================================

When a Haskell application is slow or uses too much memory,

> Cabal and the
> `Haskell GHC compiler <https://downloads.haskell.org/ghc/latest/docs/users_guide/>`__

Reviewer: ```suggestion
Cabal and `GHC <https://downloads.haskell.org/ghc/latest/docs/users_guide/>`__
```

can help you understand why.

The main steps are to

1. let GHC insert performance measuring code into your application,
2. run the application to collect a performance report and
3. visualize that report.

The process of inserting performance measuring code and collecting performance information
is called "profiling". Cabal makes this easy to do.

Profiling CPU performance
-------------------------

First, build your application, e.g. ``my-app``, with profiling enabled:

.. code-block:: console

    $ cabal build --enable-profiling --profiling-detail=late my-app

Under the hood, this will pass the corresponding
`profiling compiler options <https://downloads.haskell.org/ghc/latest/docs/users_guide/profiling.html#compiler-options-for-profiling>`__

> to GHC. The additional Cabal flag ``--profiling-detail=late`` instructs GHC to use

Reviewer 2: There should be a mention of compatible GHC versions. Late appeared only in 9.10, I think. And, for instance, ghcup still recommends 9.4.

Reviewer 3: It was introduced in 9.4 (https://downloads.haskell.org/ghc/9.4.1/docs/users_guide/9.4.1-notes.html#compiler).

Reviewer 2: Oh, I mixed it up with something else then. Thanks for catching it!

I retract the comment

Reviewer 2: I’m fine with mentioning the version but I think it’s fine either way because 9.4 is pretty old by now, I feel like

so-called
`late-cost-center profiling <https://downloads.haskell.org/ghc/latest/docs/users_guide/profiling.html#ghc-flag--fprof-late>`__
and insert performance measuring code only after important optimisations
have been applied to your application code.
This reduces the performance slow-down of profiling itself and gives you more realistic measurements.
See the Cabal section on :ref:`profiling options <profiling-options>` for more details.

.. note::
    You can add further three recommended profiling options to your ``my-app.cabal`` file
    to control where measuring code is inserted into your application as follows:

    ::

        executable my-app
          ...

>           ghc-options:

Reviewer 3: This is another place where `ghc-prof-options` might make more sense.

Author: I just renamed the option to `ghc-prof-options` (i had overlooked the precise spelling). What other place do you mean? I couldn't find any.

Reviewer 3: @Kleidukos mentioned it [here](https://github.com/haskell/cabal/pull/10539#issuecomment-2465305967) although in fact I now think he was referring to this same place, not a different one.

Other report format options can be found in the
`GHC format documentation. <https://downloads.haskell.org/ghc/latest/docs/users_guide/profiling.html#time-and-allocation-profiling>`__.

.. note::

    You can combine both steps above into a single ``cabal run`` command:

    .. code-block:: console

      $ cabal run --enable-profiling --profiling-detail=late my-app -- +RTS -pj -RTS
      <program runs and finishes>

Finally, load the profiling report file  ``my-app.prof`` into a visualizer
and look for performance bottlenecks. One popular open-source
`flame graph <https://www.brendangregg.com/flamegraphs.html>`__
visualizer is
`Speedscope <https://speedscope.app>`__,
which runs in the browser and can open this report directly.

Profiling your dependencies too
-------------------------------

The setup so far only profiles your main application, which is usually what you want.
This happens by default, because Cabal command line options only apply to local packages
and dependencies are usually not local.
However, the bottlenecks may be in your dependencies, so you would want to profile those too.

> First, to enable ``late``-cost-center profiling`` of all packages/dependencies in your project,
> add the following to your project’s ``cabal.project``` file:

Reviewer 4: Dangling backticks

```suggestion
First, to enable ``late``-cost-center profiling of all packages/dependencies in your project,
add the following to your project’s ``cabal.project`` file:
```

How to analyze Haskell performance
==================================

When a Haskell application is slow or uses too much memory,

> Cabal and `GHC <https://downloads.haskell.org/ghc/latest/docs/users_guide/>`__

Reviewer 3: Should this point to the [profiling section](https://downloads.haskell.org/ghc/latest/docs/users_guide/profiling.html) of the GHC manual?

Author: done

can help you understand why.

> The main steps are to
> 
> 1. let GHC insert performance measuring code into your application,
> 2. run the application to collect a performance report and
> 3. visualize that report.

Reviewer 4: ```suggestion
The main steps are to

1. let GHC insert performance measuring code into your application (or insert them `manually <https://downloads.haskell.org/ghc/latest/docs/users_guide/profiling.html#inserting-cost-centres-by-hand>`__),
2. run the application with the right RTS flags to produce a performance report and
3. visualize and analyze that report.
```

The process of inserting performance measuring code and collecting performance information

> is called "profiling". The main work is done by GHC while Cabal only acts

Reviewer 4: ```suggestion
is called "profiling". The work is done by GHC while Cabal only acts
```

as a build configuration interface and passes corresponding compiler flags to GHC.

Profiling CPU performance
-------------------------

The first step to build your application, e.g. ``my-app``, with profiling enabled, and
the second step to run it to collect a report, can be done with a single ``cabal run`` command:

.. code-block:: console

      $ cabal run --enable-profiling --profiling-detail=late my-app -- +RTS -pj -RTS

>       <program runs and finishes>

Reviewer 4: I would rather first mention the `cabal configure` path, then in a note mention that you can do this in one go but it is not persistent.

Author: Done

>       <program runs and finishes>

Reviewer 4: I would mention that there are two outputs, in particular `-p` and `-pj` and that you will, just for educational purposes, show the `-pj` in particular.

Reviewer 4: From my comment below, I would actually mention the three if we are going to mention any at all.

Reviewer 2: Yes, please! The eventlog property that Javier is getting at is called time series, I believe? https://en.wikipedia.org/wiki/Time_series

Author: All three mentioned.

Finally, a profiling JSON report is written to a ``<app-name>.prof`` file,
i.e. ``my-app.prof``, in the current directory.
Load the profiling report file  ``my-app.prof`` into a visualizer
and look for performance bottlenecks. One popular open-source
`flame graph <https://www.brendangregg.com/flamegraphs.html>`__
visualizer is
`Speedscope <https://speedscope.app>`__,

> which runs in the browser and can open this JSON file directly.

Reviewer 4: I would mention something like:

- `-pj` produces JSON output, can be visualized in `speedscope`.
- `-p` produces GHC's own `.prof` format, can be visualized in `profiteur` or `ghcprofview`.
- `-l -p` produces an eventlog that can be visualized in `speedscope` by first converting it via `hs-speedscope`.

Reviewer 4: Also I'm not 100% sure of that I'm about to say, but my understanding is that `-p` and `-pj` shows total time spent, and even in speedscope you see totals. One can see how much time a function took, but not when.

I think the eventlog path shows totals but also shows when this happened, i.e. information will be interleaved with GCs or switching contexts, and split by capabilities. IMHO this is more useful and the thing I usually use.

Author: Added the three examples, although i think you might have meant `-l-p`, because `-l -p` (with a space inbetween) just produces two separate files `.eventlog` and `.prof`.

Reviewer 4: Added a comment in my last review. It is actually `-l -p`. `-l-p` is "eventlog without parallel sparks" which is not what we want here.

Reviewer 4: Closing this comment in favor of the other one.

See the
`Haskell Optimization Handbook <https://haskell.foundation/hs-opt-handbook.github.io>`__
on how to optimize your code based on the profiling results afterwards.

The ``cabal run`` command above is essentially a shorthand for

.. code-block:: console

    $ cabal build --enable-profiling --profiling-detail=late my-app
    $ cabal list-bin my-app
    /path/to/my-app
    $ /path/to/my-app +RTS -pj -RTS

>     <program runs and finishes>

Reviewer 4: As said, I would mention this first and then the compressed command.

Author: Done

How to analyze Haskell performance
==================================

When a Haskell application is slow or uses too much memory,
Cabal and `GHC <https://downloads.haskell.org/ghc/latest/docs/users_guide/profiling.html>`__
can help you understand why. The main steps are:

1. Configure the project in a way that makes GHC insert performance-measuring code into your application.
2. Run the application with the right
   `runtime system (RTS) flags <https://downloads.haskell.org/ghc/latest/docs/users_guide/runtime_control.html>`__
   to produce a performance report and
3. Visualize and analyze that report.

The process of inserting performance measuring code and collecting performance information
is called "profiling". The work is done by GHC while Cabal only acts
as a build configuration interface and passes corresponding compiler flags to GHC.

Profiling CPU performance
-------------------------

First, configure Cabal to build your application, e.g. ``my-app``, with profiling enabled,
with the following command:

.. code-block:: console

    $ cabal configure --enable-profiling

> This command creates a `cabal.project.local <cabal-project-file>`__ file with the following content:

Author: Fixed

is written to a ``<app-name>.prof`` file, i.e. ``my-app.prof``, in the current directory.

.. note::

    Different report formats can be generated by using different RTS flags. Some useful ones are:

    - ``-p`` for a GHC's own
      `standard report <https://downloads.haskell.org/ghc/latest/docs/users_guide/profiling.html#cost-centres-and-cost-centre-stacks>`__
      ``<app-name>.prof``, which can be visualized with `profiteur <https://github.com/jaspervdj/profiteur>`__
      or `ghcprofview <https://github.com/portnov/ghcprofview-hs>`__.
    - ``-pj`` for a
      `JSON report <https://downloads.haskell.org/ghc/latest/docs/users_guide/profiling.html#json-profile-format>`__
      ``<app-name>.prof``, which can be visualized with `Speedscope <https://speedscope.app>`__.
    - ``-l-p`` for a binary
      `"eventlog" report <https://downloads.haskell.org/ghc/latest/docs/users_guide/runtime_control.html#rts-eventlog>`__
      ``<app-name>.eventlog``, which contains a lot more details and can show you resource usage over time, and can
      be converted to JSON with `hs-speedscope <https://github.com/mpickering/hs-speedscope>`__

>       to be visualized with `Speedscope <https://speedscope.app>`__.

Author: Fixed and rephrased

>       to be visualized with `Speedscope <https://speedscope.app>`__.

Author: Fixed

Finally, visualize this JSON report ``my-app.prof`` and analyze it for performance bottlenecks.
One popular open-source
`flame graph <https://www.brendangregg.com/flamegraphs.html>`__
visualizer is
`Speedscope <https://speedscope.app>`__,
which runs in the browser and can open this JSON file directly.
See the
`Haskell Optimization Handbook <https://haskell.foundation/hs-opt-handbook.github.io>`__
on how to optimize your code based on the profiling results afterwards.

So far, we've only used a single GHC compiler option to enable profiling
in general for your application.
However, you can pass a lot more compiler flags to GHC via the ``cabal.project.local`` file:

.. code-block:: cabal

    profiling: True
    profiling-detail: late-toplevel
    program-options
      ghc-options:
        -fprof-auto
        -fno-prof-count-entries

>         -fprof-auto-calls

Reviewer 4: There is a bit of confusion here. `-fprof-auto` is the same as `profiling-detail: all-functions`, and `-fprof-auto-calls` is another one of these flags so you are specifying a (kind of) profiling detail thrice. I assume maybe they all are applied and you get all non-inline definitions + all call sites + all top-level after optimizations?

Also if you mention program-options, I would link to the appropriate place: https://cabal.readthedocs.io/en/latest/cabal-project-description-file.html#cfg-section-program-options-program-options

Reviewer 4: It would be nice if the code blocks could contain links. I don't know if that is possible in RST.

Reviewer 3: I don't think it is: the main point of a code block is to disable markup which might have a different meaning in code.

Author: I added a sentence about `program-options` below with a reference link.

Regarding those specific `ghc-options` i added them, because @Kleidukos suggested them (i can imagine these are a good default to promote), but i have never used them before. I added the link to the GHC explanations a bit below, which i wouldn't want to duplicate here.

Reviewer 4: I'm unsure, I think `-fprof-auto` and `-fprof-auto-calls` will get in the way for optimizations, which is precisely what we would like to avoid with `-fprof-late`. I don't think these should be advertised together.

Reviewer 5: I stand corrected. So, in context of late profiling, `-fprof-auto` just extends it beyond top-level definitions? If so, it makes sense to include it, especially if the Cabal manual is correct that our `late` is equivalent to `late-toplevel`, so there's no easy way to demand more than top-level definitions. I couldn't find in the cabal manual whether late or non-late profiling is the cabal default (GHC's is late, I think, so maybe it's inherited from the GHC in use?). Anyway, sorry for the noise and I'm glad this gets enough pairs of eyes.

Author: So is this resolved/okay?

Reviewer 4: I think I would write:

> So far, we’ve only used a single cabal option to enable profiling in general for your application. Leaving [`profiling-detail`](https://cabal--10539.org.readthedocs.build/en/10539/cabal-project-description-file.html#cfg-field-profiling-detail) unspecified defaults to `profiling-detail: exported-functions` for libraries and `profiling-detail: toplevel-functions` for executables. 
>
> However, you can specify the profiling detail and provide more compiler flags to GHC (such as `-fno-prof-count-entries`) via the `cabal.project.local` file:
> ```
> profiling: True
> profiling-detail: late-toplevel
> program-options
>   ghc-options:
>     <further options>
> ```

Reviewer 5: Sounds good, but I'd double check that we indeed did not follow GHC and did not make late profiling the default. Or maybe we insert the same GHC flags but now they mean late profiling?

Author: I applied this paragraph. If it's not accurate, please provide an alternative suggestions .

Reviewer 5: I don't have a reasonable suggestion without code-diving, but fortunately others have commented, so let's focus there.  Thank you!

Add these recommended profiling options to control where and when

> performance measuring code is inserted into your application

Reviewer 4: ```suggestion
performance measuring code is inserted into your application.
```

The setting ``profiling-detail: late-toplevel`` instructs GHC to use so-called
`late-cost-center profiling <https://downloads.haskell.org/ghc/latest/docs/users_guide/profiling.html#ghc-flag--fprof-late>`__
and insert measuring code only after important optimisations have been applied to your application code.
This reduces the performance slow-down of profiling itself and gives you more realistic measurements.

The ``ghc-options`` setting allows you to further control which functions and other bindings

> the GHC compiler should profile.

Reviewer 4: ```suggestion
the GHC compiler should profile or other aspects of profiling.
```

In particular `-fno-prof-count-entries` has a different purpose.

You can find more information and further options in the
`GHC "cost-center" guide <https://downloads.haskell.org/ghc/latest/docs/users_guide/profiling.html#automatically-placing-cost-centres>`__.
and the
`GHC profiling compiler options <https://downloads.haskell.org/ghc/latest/docs/users_guide/profiling.html#compiler-options-for-profiling>`__
section.

Profiling your dependencies too
-------------------------------

The setup so far only profiles your main application, which is usually what you want.

> This happens by default, because Cabal command line options only apply to local packages

Author: Rephrased

and dependencies are usually not local.
However, the bottlenecks may be in your dependencies, so you would want to profile those too.

> First, to enable ``late``-cost-center profiling of all packages/dependencies in your project,

Reviewer 2: ```suggestion
First, to enable ``late``-cost-center profiling of all packages (including dependencies) concerning your project,
```

Author: Fixed

add the following to your project’s ``cabal.project.local`` file:

.. code-block:: cabal

    package *
        profiling: true

>         profiling-detail: late

Reviewer 4: As you mentioned `program-options` above, you can mention here that further options might be passed via:
```
package *
  profiling: true
  profiling-detail: late
  ghc-options:
    -fno-prof-count-entries
```

Reviewer 4: Also perhaps I would mention that one should add:
```
package <my-dependency>
  profiling: true
  profiling-detail: late
  ghc-options:
    -fno-prof-count-entries
```

to the cabal.project.local, and *if one wants it to apply to all dependencies* then use `package *`. See the program-options section https://cabal.readthedocs.io/en/latest/cabal-project-description-file.html#cfg-section-program-options-program-options

- `program-options`: all local packages
- `package blah`: only package `blah` (be it local or remote)
- `package *`: all local + remote packages

Reviewer 4: This is now a bit wrong. `package <package-name` applies to a package, be it local or remote.

Author: Fixed

How to analyze Haskell performance
==================================

When a Haskell application is slow or uses too much memory,
Cabal and `GHC <https://downloads.haskell.org/ghc/latest/docs/users_guide/profiling.html>`__
can help you understand why. The main steps are:

1. Configure the project in a way that makes GHC insert performance-measuring code into your application.
2. Run the application with the right
   `runtime system (RTS) flags <https://downloads.haskell.org/ghc/latest/docs/users_guide/runtime_control.html>`__
   to produce a performance report and
3. Visualize and analyze that report.

The process of inserting performance measuring code and collecting performance information
is called "profiling". The work is done by GHC while Cabal only acts
as a build configuration interface and passes corresponding compiler flags to GHC.

> 

Reviewer 4: Or something along those lines

`GHC "cost-center" guide <https://downloads.haskell.org/ghc/latest/docs/users_guide/profiling.html#automatically-placing-cost-centres>`__.
and the
`GHC profiling compiler options <https://downloads.haskell.org/ghc/latest/docs/users_guide/profiling.html#compiler-options-for-profiling>`__
section.

Profiling your dependencies too
-------------------------------

The setup so far only profiles your main application, which is usually what you want.
This happens by default, because Cabal command line options only apply to local packages
and dependencies are usually not local.
However, the bottlenecks may be in your dependencies, so you would want to profile those too.

First, to enable ``late``-cost-center profiling for all packages (including dependencies) in your project,
not just the local ones, add the following to your project’s ``cabal.project.local`` file:

.. code-block:: cabal

    package *
        profiling: true
        profiling-detail: late

.. note::

    There are several keywords to specify to which parts of you project settings should be applied:

    - ``program-options`` to apply to :ref:`all local packages<program_options>`.

>     - ``package <package-name>`` to apply to a :ref:`single, specific local package<package-configuration-options>`.

Reviewer 2: ```suggestion
    - ``package <package-name>`` to apply to a :ref:`single package<package-configuration-options>`, weather local or remote.
```

   to produce a performance report.
3. Visualize and analyze that report.

The process of inserting performance measuring code and collecting performance information
is called "profiling".
This guide describes how to instruct Cabal to pass desired profiling flags to the GHC compiler;
Cabal acts as a convenient build configuration interface while the work is done by GHC.
To get a deeper understanding of the overall profiling process itself in GHC,
it is highly recommended to read in depth the
`Profiling section in GHC's User Guide <https://downloads.haskell.org/ghc/latest/docs/users_guide/profiling.html>`__.

Profiling CPU performance
-------------------------

First, configure Cabal to build your application, e.g. ``my-app``, with profiling enabled,
with the following command:

.. code-block:: console

    $ cabal configure --enable-profiling

This command creates a ``cabal.project.local`` file with the following content:

.. code-block:: cabal

      profiling: True

This file stores temporary configuration settings that are passed implicitly to further Cabal commands
like ``cabal build`` and ``cabal run``.
The setting ``profiling: True`` tells GHC

> to insert performance measuring code into your application in general,

Author: Added

packages of your project (See :ref:`Program options<program_options>`).
The ``ghc-options`` setting allows you to further control which functions and other bindings
the GHC compiler should profile, as well as other aspects of profiling.
You can find more information and further options in the
`GHC "cost-center" guide <https://downloads.haskell.org/ghc/latest/docs/users_guide/profiling.html#automatically-placing-cost-centres>`__.
and the
`GHC profiling compiler options <https://downloads.haskell.org/ghc/latest/docs/users_guide/profiling.html#compiler-options-for-profiling>`__
section.

Profiling your dependencies too
-------------------------------

The profiling setup so far with the ``cabal.project.local`` file only applied to your local packages,
which is usually what you want.
However, bottlenecks may also exist in your dependencies, so you may want to profile those too.

First, to enable ``late``-cost-center profiling for all packages (including dependencies) concerning your project,
not just the local ones, add the following to your project’s ``cabal.project.local`` file:

.. code-block:: cabal

    package *

>         profiling: true

Reviewer 3: I'm pretty sure I mentioned that earlier. (Yep: https://github.com/haskell/cabal/pull/10539#discussion_r1834865686)

>         profiling: true

Reviewer 3: Straight from the horse's mouth: https://downloads.haskell.org/ghc/latest/docs/users_guide/profiling.html#compiler-options-for-profiling

> To make use of the profiling system _all_ modules must be compiled and linked with the `-prof` option. Any `SCC` annotations you’ve put in your source will spring to life.

(Emphasis theirs.)

How to analyze Haskell performance
==================================

When a Haskell application is slow or uses too much memory,
Cabal and `GHC <https://downloads.haskell.org/ghc/latest/docs/users_guide/profiling.html>`__
can help you understand why.

The main steps are to

> 1. let GHC insert performance measuring code into your application (or insert them
>    `manually <https://downloads.haskell.org/ghc/latest/docs/users_guide/profiling.html#inserting-cost-centres-by-hand>`__),
> 2. run the application with the right
>    `runtime system (RTS) flags <https://downloads.haskell.org/ghc/latest/docs/users_guide/runtime_control.html>`__
>    to produce a performance report and
> 3. visualize and analyze that report.

Reviewer 2: These three steps, ideally, should serve as an outline to what follows. So, I'd suggest to rephrase the first one a little bit to mention "configuring", and the second one will map ideally on `cabal run` once you introduce it.

Also, the `N.`-lists should use sentence case (capital letters at the start of every point).

```suggestion
1. Configure the project in a way that makes GHC insert performance-measuring code into your application (or insert them.
2. Run the application with the right
   `runtime system (RTS) flags <https://downloads.haskell.org/ghc/latest/docs/users_guide/runtime_control.html>`__.
   to produce a performance report and
3. Visualize and analyze that report.
```

And what follows below should mention "step one", "step two", "step three", hopefully.

One more thing: I don't think mentioning manual insertion of cost centers should be done here. It's a detail that I'd save for a later time. Hence, I remove it from here.

See the
`Haskell Optimization Handbook <https://haskell.foundation/hs-opt-handbook.github.io>`__
on how to optimize your code based on the profiling results afterwards.

Different report formats can be generated by using different RTS flags. Some useful ones are:

- ``-p`` for a GHC's own
  `standard report <https://downloads.haskell.org/ghc/latest/docs/users_guide/profiling.html#cost-centres-and-cost-centre-stacks>`__
  ``<app-name>.prof``, which can be visualized with `profiteur <https://github.com/jaspervdj/profiteur>`__
  or `ghcprofview <https://github.com/portnov/ghcprofview-hs>`__.
- ``-pj`` for a
  `JSON report <https://downloads.haskell.org/ghc/latest/docs/users_guide/profiling.html#json-profile-format>`__
  ``<app-name>.prof``, which can be visualized with `Speedscope <https://speedscope.app>`__.
- ``-l-p`` for a binary
  `"eventlog" report <https://downloads.haskell.org/ghc/latest/docs/users_guide/runtime_control.html#rts-eventlog>`__
  ``<app-name>.eventlog``, which contains a lot more details and can show you resource usage over time, and can
  be converted to JSON with `hs-speedscope <https://github.com/mpickering/hs-speedscope>`__

>   and visualized with `Speedscope <https://speedscope.app>`__.

Reviewer 2: GHC manual mentions only ThredScope. I wonder if it's a good idea to mention a workflow that doesn't seem to have an official blessing. But on the other hand, I much prefer Speedscope to ThreadScope myself.

Author: Maybe the GHC doc is just not uptodate. i would like to keep mentioning speedscope.

Reviewer 5: Speedscope and ThreadScope have very different goals and use cases. Also, Threadscope can't consume JSON files. But, yes, let's make sure the GHC and cabal docs are in accord.

Reviewer 2: > Speedscope and ThreadScope have very different goals and use cases

@Mikolaj I'm not making stuff up, I swear! ThreadScope is what GHC manual suggests: https://downloads.haskell.org/ghc/latest/docs/users_guide/runtime_control.html#tracing Should we file an issue?

Reviewer 5: I think that section suggests ThreadScope for events related to threads. Eventlogs have been repurposed also for events related to profiling and, additionally, they can be output in JSON and read into Speedscope. That's the bit relevant for profiling. Yes, let's file an issue, because this section seems outdated and based on the assumption the original functionality of eventlogs is the only one that exists.

.. note::

    The report generated from the ``cabal configure``, ``cabal build`` etc. commands can
    also be generated with a single ``cabal run`` command, but no settings are persisted
    to a file:

    .. code-block:: console

          $ cabal run --enable-profiling --profiling-detail=late my-app -- +RTS -pj -RTS

>           <program builds, runs and finishes>

Author: Streamlined

You can also pass a lot more
`profiling compiler options <https://downloads.haskell.org/ghc/latest/docs/users_guide/profiling.html#compiler-options-for-profiling>`__
to GHC via the ``ghc-prof-options`` field in your ``my-app.cabal`` file.
Add the following, recommended profiling options to your ``my-app.cabal`` file
to control where performance measuring code is inserted into your application:

::

    executable my-app
      ...
      ghc-prof-options:
        -fprof-auto
        -fno-prof-count-entries
        -fprof-auto-calls

>         ...

Reviewer 2: Once again I ask to not put anything profile-related in the cabal file --- that's not what cabal files are for. These options can be added to a project(.local) file. Do you need help with figuring out how?

Author: Done. The only way i could find to pass such flags was "program options" (all in all a very confusing api). If that's not the intended name, please correct me.

Reviewer 2: Looks good! I think it may be possible to do it per `package` stanza. But I haven't checked. 

I agree it's not super clean. But we should strive to maintain the distinction between project-level options and package descriptions. Even if they are technically not orthogonal and the same could be achieved with either means.

You can find more information and further options in the
`GHC "cost-center" guide <https://downloads.haskell.org/ghc/latest/docs/users_guide/profiling.html#automatically-placing-cost-centres>`__.

Profiling your dependencies too
-------------------------------

The setup so far only profiles your main application, which is usually what you want.
This happens by default, because Cabal command line options only apply to local packages
and dependencies are usually not local.
However, the bottlenecks may be in your dependencies, so you would want to profile those too.

First, to enable ``late``-cost-center profiling of all packages/dependencies in your project,

> add the following to your project’s ``cabal.project`` file:

Reviewer 2: Should it keep using `cabal.project.local`? I think one file is a better idea than two. Especially given the argument you provide when introducing `.local`: `.project` are meant for Gitting, while `.local` are not.

Author: Let's keep it simple with a single file.
Btw, i couldn't find any mentioning of `cabal.prof.project` xD.

Reviewer 2: > Let's keep it simple with a single file.

So, `cabal.projct.local`, right? I still see `cabal.project`...

Author: Fixed

How to analyze Haskell performance
==================================

When a Haskell application is slow or uses too much memory,
Cabal and the
`Haskell GHC compiler <https://downloads.haskell.org/ghc/latest/docs/users_guide/>`__
can help you understand why.

The main steps are to
1. let GHC insert performance measuring code into your application,
2. run the application to collect a performance report and
3. visualize that report.

The process of inserting performance measuring code and collecting performance information

> is called "profiling" and Cabal makes this is easy to do:

Reviewer 3: ```suggestion
is called "profiling". Cabal makes this easy to do.
```

Profiling CPU performance
-------------------------

First, build your application, e.g. ``my-app``, with profiling enabled:

.. code-block:: console

    $ cabal build --enable-profiling --profiling-detail=late exe:my-app

Under the hood this will pass the corresponding
`profiling compiler options <https://downloads.haskell.org/ghc/latest/docs/users_guide/profiling.html#compiler-options-for-profiling>`__

> to GHC. With the additional Cabal flag ``--profiling-detail=late`` GHC is instructed to use

Reviewer 3: ```suggestion
to GHC. The additional Cabal flag ``--profiling-detail=late`` instructs GHC to use
```

`"late-cost-center" profiling <https://downloads.haskell.org/ghc/latest/docs/users_guide/profiling.html#ghc-flag--fprof-late>`__
and insert performance measuring code only after important optimisations

> have been applied to minimize the impact of profiling on performance.

Reviewer 3: ```suggestion
have been applied, to minimize the impact of profiling on performance.
```
(At "to", it could either say what it's applying optimizations to, or it could be explaining. This seems clearer for non-native English speakers.)

See the Cabal section on :ref:`profiling options <profiling-options>` for more details.

Second, run the application with the necessary
`runtime system (RTS) options <https://downloads.haskell.org/ghc/latest/docs/users_guide/runtime_control.html>`__
to produce a profile report.

> .. code-block:: console

Reviewer 3: Do we want to suggest `$(cabal list-bin my-app) +RTS -pj -RTS` instead of needing to copy the path? P personally use that idiom fairly often when testing changes.

Author: Good idea, added.

    $ cabal list-bin exe:my-app
    /path/to/my-app
    $ /path/to/my-app +RTS -pj -RTS
    <program runs and finishes>

The report is written to a ``<app-name>.prof`` file, i.e. ``my-app.prof``, in the current directory.
With the RTS option ``-pj`` the app produces a
`"profile JSON" (pj) file report <https://downloads.haskell.org/ghc/latest/docs/users_guide/profiling.html#rts-flag--pj>`__.
Other report format options can be found in the
`GHC format documentation. <https://downloads.haskell.org/ghc/latest/docs/users_guide/profiling.html#time-and-allocation-profiling>`__.

Finally, load the profiling report file into a visualizer to look for performance bottlenecks.
One popular open-source
`flame graph <https://www.brendangregg.com/flamegraphs.html>`__
visualizer is `Speedscope <https://speedscope.app>`, which runs in the browser and comes with
an example.

Profiling your dependencies too
-------------------------------

The setup so far only profiles your main application, which is usually what you want.
This happens by default, because Cabal command line options only apply to local packages
and dependencies are usually not local.

> However, the bottlenecks may be in your dependencies so you would want to profile those too.

Reviewer 3: ```suggestion
However, the bottlenecks may be in your dependencies, so you would want to profile those too.
```

Reviewer 3: Also, doesn't GHC require profiled builds of dependencies to go along with the profiled build of your application (see compiler ways)?

Reviewer 6: > Also, doesn't GHC require profiled builds of dependencies to go along with the profiled build of your application

AFAIU, it requires the dependencies to be compiled with `-prof`, but you don't need to insert any SCCs there if you don't want to.

Reviewer 6: FWIW, GHC manual is not very explicit about that, in https://downloads.haskell.org/ghc/9.10-latest/docs/users_guide/profiling.html it says

> Re-compile your program for profiling with the -prof option, and probably one of the options for adding automatic annotations: -fprof-late is the recommended option.

That implies that you need to recompiled everything; but I cannot find any explicit mention that mixing (i.e. linking) `-prof` and non-`-prof` objects would not work. 

It wont, you may get a compile time error like

```
ghc -c Bar.hs      

Bar.hs:3:1: error:
    Bad interface file: ./Foo.hi
        mismatched interface file profile tag (wanted "", got "p")
  |
3 | import Foo
  | ^^
```

but again, it's not mentioned in the documentation. (Though probably only cabal developers would be interested in such sort of information, I think it's good to exist).

Reviewer 3: I think the closest it gets is the manual quote I put later (https://github.com/haskell/cabal/pull/10539#discussion_r1855251873). But that could be taken to mean that profiling would be ineffective, as opposed to causing compilation failure due to mismatched ways.

Reviewer 7: > but I cannot find any explicit mention that mixing (i.e. linking) -prof and non--prof objects would not work.

I guess it's true that technically the way it's phrased does not specify this. If someone is willing to open a ticket suggesting a more accurate phrasing on the ghc issue tracker I can take care of the actual MR. Just assign me to the ticket if you do.

Reviewer 3: https://gitlab.haskell.org/ghc/ghc/-/issues/25518 and I've assigned you (assuming you're AndreasK on GitLab).

> To enable ``late``-cost-center profiling`` all packages/dependencies in your project,

Reviewer 3: ```suggestion
To enable ``late``-cost-center profiling`` of all packages/dependencies in your project,
```

add the following to your project’s ``cabal.project``` file:

.. code-block:: console

    package *
        profiling: true
        profiling-detail: late

> So, rebuild your application with ``cabal build``:

Reviewer 3: ```suggestion
Then rebuild your application with ``cabal build``:
```
